spec/src/consensus.rs

```diff
@@ -195,15 +199,64 @@ impl Consensus {
             block_version: BLOCK_VERSION,
             proposer_reward_ratio: PROPOSER_REWARD_RATIO,
             max_block_proposals_limit: MAX_BLOCK_PROPOSALS_LIMIT,
+            satoshi_pubkey_hash: SATOSHI_PUBKEY_HASH,
+            satoshi_cell_occupied_ratio: SATOSHI_CELL_OCCUPIED_RATIO,
         }
     }
 
-    pub fn set_id(mut self, id: String) -> Self {
+    pub fn build(self) -> Consensus {
+        let ConsensusBuilder {
+            genesis_hash,
+            genesis_block,
+            id,
+            max_uncles_num,
+            epoch_reward,
+            orphan_rate_target,
+            epoch_duration_target,
+            secondary_epoch_reward,
+            tx_proposal_window,
+            pow,
+            cellbase_maturity,
+            median_time_block_count,
+            max_block_cycles,
+            max_block_bytes,
+            genesis_epoch_ext,
+            block_version,
+            proposer_reward_ratio,
+            max_block_proposals_limit,
+            satoshi_pubkey_hash,
+            satoshi_cell_occupied_ratio,
+        } = self;
+        Consensus {
+            genesis_hash,
+            genesis_block,
+            id,
+            max_uncles_num,
+            epoch_reward,
+            orphan_rate_target,
+            epoch_duration_target,
+            secondary_epoch_reward,
+            tx_proposal_window,
+            pow,
+            cellbase_maturity,
+            median_time_block_count,
+            max_block_cycles,
+            max_block_bytes,
+            genesis_epoch_ext,
+            block_version,
+            proposer_reward_ratio,
+            max_block_proposals_limit,
+            satoshi_pubkey_hash,
+            satoshi_cell_occupied_ratio,
+        }
+    }
+
+    pub fn id(mut self, id: String) -> Self {
         self.id = id;
         self
     }
 
-    pub fn set_genesis_block(mut self, genesis_block: BlockView) -> Self {
+    pub fn genesis_block(mut self, genesis_block: BlockView) -> Self {
```

Comment:
I think this builder didn't fix the issue that why we need a builder, we still use setters with side effects.

```rust
pub fn genesis_block(mut self, genesis_block: BlockView) -> Self {
    // ... omit ...
    self.genesis_epoch_ext
        .set_difficulty(genesis_block.difficulty());	            .set_difficulty(genesis_block.difficulty());
    self.genesis_hash = genesis_block.hash();	        self.genesis_hash = genesis_block.hash();
    // ... omit ...
}
```

Anyway, we can do this optimization later.

Reply:
I am curious how can a setter can without side effects?